backend/executor.go

```diff
@@ -202,22 +228,76 @@ func (g *grpcExecutor) GetWorkItems(req *protos.GetWorkItemsRequest, stream prot
 	callback := g.onWorkItemConnection
 	if callback != nil {
 		if err := callback(stream.Context()); err != nil {
-			message := fmt.Sprint("unable to establish work item stream at this time: ", err)
+			message := "unable to establish work item stream at this time: " + err.Error()
 			g.logger.Warn(message)
 			return status.Errorf(codes.Unavailable, message)
 		}
 	}
 
+	// Collect all pending activities on this stream
+	// Note: we don't need sync.Map's here because access is only on this thread
+	pendingActivities := make(map[string]struct{})
+	pendingActivityCh := make(chan string, 1)
+	pendingOrchestrators := make(map[string]struct{})
+	pendingOrchestratorCh := make(chan string, 1)
+	defer func() {
+		// If there's any pending activity left, remove them
+		for key := range pendingActivities {
```

Comment:
What is the use of `pendingActivities` or `pendingOrchestrators` here? Since we just need to delete all pending keys that are there in `g.pendingActivities` or `g.pendingOrchestrators`, can we not simply iterate in the map and delete? This would avoid sending the items to pendingChannel as well for every item being processed..

Reply:
I think there are at least two reasons to use `pendingActivities` and `pendingOrchestrators`
1. they are not `sync.Map`, they should have possibly better performance. 
2. they are empty maps, which won't use much space when there are lots of pending orchestrators/activities.

Reply:
No, we do not delete all keys in g.pendingActivities.

There could be multiple connections to GetWorkItems, so each connection needs its own map to track what pending items are related to that connection.

Using a local variable, rather than depending on the global g.pendingActivities/g.pendingOrchestrators, means we can avoid using locks (because of how gRPC works, there's no issue with concurrent access there) and simplify the cleanup process.

Reply:
Oh interesting. We support multiple connections to GetWorkItems? I didn't know and was actually going to open an issue for it! 

A question though: In case we have say 2 connections to the server, when the workItem gets added to `g.workItemQueue`, how does it properly gets allocated to the correct stream? It seems like two concurrent streams trying to read a common channel.. won't that have any issue?

Reply:
Got it. In that case, it's required to maintain the stream local map. Thanks for explaining.